src/shared/components/common/emoji-picker.tsx

```diff
@@ -0,0 +1,99 @@
+import { Component } from "inferno";
+import { i18n } from "../../i18next";
+import { getEmojiMart } from "../../utils";
+import { Icon } from "./icon";
+
+
+interface EmojiMartProps {
+    onEmojiClick?(val: any): any;
+    pickerOptions: any;
+}
+
+interface EmojiMartState {
+}
+
+export class EmojiMart extends Component<
+    EmojiMartProps,
+    EmojiMartState
+> {
+    private emptyState: EmojiMartState = {
+    };
+    constructor(props: any, context: any) {
+        super(props, context);
+        this.state = this.emptyState;
+    }
+    componentDidMount() {
+        let div: any = document.getElementById("emoji-picker");
+        if (div) {
+            div.appendChild(getEmojiMart(this.handleEmojiClick,this.props.pickerOptions));
+        }
+    }
+
+    render() {
+        return (
+            <div id="emoji-picker"></div>
+        );
+    }
+
+    handleEmojiClick = (e: any) => {
+        if (this.props.onEmojiClick)
+            this.props.onEmojiClick(e);
+    }
+}
+
+interface EmojiPickerProps {
+    onEmojiClick?(val: any): any;
+}
+
+interface EmojiPickerState {
+    showPicker: boolean,
+}
+
+export class EmojiPicker extends Component<
+    EmojiPickerProps,
+    EmojiPickerState
+> {
+    private emptyState: EmojiPickerState = {
+        showPicker: false,
+    };
+    constructor(props: any, context: any) {
+        super(props, context);
+        this.state = this.emptyState;
+    }
+    render() {
+        return (
+            <span>
+                <button
+                    className="btn btn-sm text-muted"
+                    data-tippy-content={i18n.t("emoji")}
+                    aria-label={i18n.t("emoji")}
+                    onClick={this.togglePicker}
+                >
+                    <Icon icon="smile" classes="icon-inline" />
+                </button>
+
+                {this.state!.showPicker && (
```

Comment:
What's the exclamation point do here?

Reply:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator

asserts that object is not null/undefined to the compiler.

Reply:
noting also, i've fixed most of these anyways